### PR TITLE
Multivariable Taylor Polynomials

### PR DESCRIPTION
**Pull request by ryyanmapes**
Adds https://github.com/asc-community/AngouriMath/issues/436

### CODE REVIEW

File: Sources/AngouriMath/Functions/Series.cs [hunk 1]
```diff
@@ -44,5 +45,100 @@ internal static Entity TaylorExpansion(Entity expr, Variable exprVariable, Varia
...
+                                    break;
+                                }
+                            }
+
+                            // If the partial derivative has already been computed, we just add to it's coefficient instead of duplicating it.
+                            if (!foundRepeatFlag)
+                            {
+                                var newPartialDerivative = lastTerm.partialDerivative.Differentiate(triple.exprVariable);
+                                newTerms.Add((lastTerm.coefficient, newPointCoeffs, newPartialDerivative));
+                            }
```
Comment (WhiteBlackGoose) [inline Sources/AngouriMath/Functions/Series.cs:95]: Okay so is it to avoid computating `Fxy` when `Fyx` is already computed, right? Could you add a few comments explaning this chunk of code?

Also, I might be missing it, but do you multiply each partial derivative according to the coefficient, like for example when you need to multiply Fxy by two in `Fxx / 2 ... + 2Fxy / 2 ... + Fyy / 2...`? If yes, please somehow minimally explain it

**Review by WhiteBlackGoose — COMMENTED** [review]
First of all, great job! It's quite a lot of work done here. A lot of useful work.

There are a few minor points to fix. Also, the algorithm is definitely not tiny, so it would be nice to supply it with more comments, so that it would be easy to understand how it works if we need to fix/rewrite it in the future.

**Comment by ryyanmapes** [comment]
Thank you so much for the feedback! Sorry about how much code was needed for this task, I'm sure there's a more readable way of doing it, but I'm not well-versed in the tools the library provides yet. I think I fixed up everything you mentioned though.

File: Sources/Tests/UnitTests/Calculus/SeriesTest.cs [hunk 2]
```diff
@@ -19,6 +20,24 @@ public void TaylorDirect(string funcOverTRaw, string expectedRaw, int termCount)
             Assert.Equal(expected, actual);
         }
 
+        [Theory]
+        [InlineData("a t3 + b t2 + c t + d", "((d) + ((c) * (x))) + (((((2) * (b)) * ((x) ^ (2))) / ((2)!)) + ((((6) * (a)) * ((x) ^ (3))) / ((3)!)))", 4)]
+        [InlineData("(e^t)ln(1+s)", "((0) + ((y) + ((((x) * (y)) + -((y) ^ (2))) / ((2)!))))", 3)]
+        [InlineData("sin(t)+cos(s)+tan(u)", "((1) + ((x) + (z)))", 2)]
```
Comment (WhiteBlackGoose) [inline Sources/Tests/UnitTests/Calculus/SeriesTest.cs:26]: To be honest... that's still too few tests. This is a big thing (multivariable taylor), it should be tested properly. Could you add tests for, like, 4 terms? Also, for single variable? For two? For four? For one term? Etc.

File: Sources/AngouriMath/Functions/Series.cs [hunk 3]
```diff
@@ -44,5 +45,110 @@ internal static Entity TaylorExpansion(Entity expr, Variable exprVariable, Varia
...
+                    fullExpr /= ((Entity)order).Factorial();
+
+                yield return fullExpr;
+
+                lastTerms = newTerms;
+                order++;
+            }
+        }
+
+        internal static Entity MultivariableTaylorExpansion(Entity expr, (Variable exprVariable, Variable polyVariable, Entity point)[] exprToPolyVars, int termCount)
```
Comment (WhiteBlackGoose) [inline Sources/AngouriMath/Functions/Series.cs:140]: Also, could you slightly change the API? To be precise, move `termCount` after `expr` and mark the array with `params` (so that we don't have to write a lot of unnecessary code, you know)

**Comment by ryyanmapes** [comment]
It looks like there was an issue in the 'CheckTheCorrectness' unit tests, where an int was being fed in as the point to check instead of a double, so I've fixed that too.

File: Sources/Tests/UnitTests/Calculus/SeriesTest.cs [hunk 4]
```diff
@@ -52,5 +75,29 @@ public void CheckTheCorrectness(double point, int termCount, string func, int po
...
             Assert.True(error < 0.05, $"Error is: {error}");
         }
+
+        [Theory]
+        [InlineData(new double[3] { 0, 0, 0 }, 4, "sin(x) + cos(y) + tan(z)", new double[3] { 0.5, -0.5, 0.1 })]
+        [InlineData(new double[3] { 0, 0, 0 }, 4, "sqrt(x - 1) / e ^ (y - 1) + sin(z)", new double[3] { 0.1, 0.1, 0.1 })]
+        [InlineData(new double[3] { 1, 1, 1 }, 6, "(e^x)ln(1 + y)z", new double[3] { 0.5, 0.5, 0.5 })]
+        public void CheckMultivariableCorrectness(double[] point, int termCount, string func, double[] pointToCheckAt)
+        {
+            var vars = new (Variable exprVariable, Variable polyVariable, Entity value)[]
```
Comment (WhiteBlackGoose) [inline Sources/Tests/UnitTests/Calculus/SeriesTest.cs:85]: We now have `params`, so you can write
```cs
MultivariableTaylorExpansion(expr, termCount, 
    ("x", "x", point[0]), ("y", "y", point[1]), ("z", "z", points[2]))
```

File: Sources/Tests/UnitTests/Calculus/SeriesTest.cs [hunk 5]
```diff
@@ -52,5 +75,29 @@ public void CheckTheCorrectness(double point, int termCount, string func, int po
 
             Assert.True(error < 0.05, $"Error is: {error}");
         }
+
+        [Theory]
+        [InlineData(new double[3] { 0, 0, 0 }, 4, "sin(x) + cos(y) + tan(z)", new double[3] { 0.5, -0.5, 0.1 })]
+        [InlineData(new double[3] { 0, 0, 0 }, 4, "sqrt(x - 1) / e ^ (y - 1) + sin(z)", new double[3] { 0.1, 0.1, 0.1 })]
+        [InlineData(new double[3] { 1, 1, 1 }, 6, "(e^x)ln(1 + y)z", new double[3] { 0.5, 0.5, 0.5 })]
+        public void CheckMultivariableCorrectness(double[] point, int termCount, string func, double[] pointToCheckAt)
```
Comment (WhiteBlackGoose) [inline Sources/Tests/UnitTests/Calculus/SeriesTest.cs:83]: Can we have more tests for non-zero point of differentiating? We have just one test. It would be nice to have tests for points with different coordinates, like (1, 2, 3) or (5, 0 -4).

File: Sources/Tests/UnitTests/Calculus/SeriesTest.cs [hunk 6]
```diff
@@ -19,6 +21,27 @@ public void TaylorDirect(string funcOverTRaw, string expectedRaw, int termCount)
             Assert.Equal(expected, actual);
         }
 
+        [Theory]
+        [InlineData("a t3 + b t2 + c t + d", "((d) + ((c) * (x))) + (((((2) * (b)) * ((x) ^ (2))) / ((2)!)) + ((((6) * (a)) * ((x) ^ (3))) / ((3)!)))", 4)]
+        [InlineData("(e^t)ln(1+s)", "(0) + ((y) + ((((2) * ((x) * (y))) + ((-1) * ((y) ^ (2)))) / ((2)!)))", 3)]
+        [InlineData("sin(t)+cos(s)+tan(u)", "((1) + ((x) + (z)))", 2)]
+        [InlineData("cos(t)sin(s)", "((0) + (y)) + ((0) + ((((-3) * (((x) ^ (2)) * (y))) + ((-1) * ((y) ^ (3)))) / ((3)!)))", 4)]
+        [InlineData("sin(t)sin(s)sin(u)sin(v)", "((0) + (0)) + ((0) + ((0) + (((24) * ((((x) * (y)) * (z)) * (w))) / ((4)!))))", 5)]
+        public void MultivariableTaylorDirect(string funcOverTRaw, string expectedRaw, int termCount)
```
Comment (WhiteBlackGoose) [inline Sources/Tests/UnitTests/Calculus/SeriesTest.cs:30]: See the previous comment, we still need more tests. For single variable you can copy the tests from single-variable taylor expansion. A one-term test should be trivial

**Review by WhiteBlackGoose — APPROVED** [review]
Looks good. I'll take a few more minutes, and if nothing in my mind changes, I'll merge.

**Comment by WhiteBlackGoose** [comment]
@ryyanmapes thank you for your contribution 🎉! You did a great job.
